### ISSUE TITLE
Is it possible to bind an object directly?

### ISSUE THREAD

**Issue by hkvstore** [BorisMoore/jsviews#427]
I tried the follows:

```html
<script id="myTmpl" type="text/x-jsrender">
  Name (as JSON): <input data-link="{toJson:name:fromJson}"/><br/><br/>

  First: <input data-link="name.first"/>{^{:name.first}}<br/>
  Last: <input data-link="name.last"/>{^{:name.last}}
</script>

<div id="page"></div>
```
```js
var myTmpl = $.templates("#myTmpl"),
  data = { name: { first: "Jo", last: "Blow" } };

$.views.converters({
  toJson: function(val) {
    return JSON.stringify(val);
  },
  fromJson: function(val) {
    return JSON.parse(val);
  }
});

myTmpl.link("#page", data);
```

When the page is run, the "Name" textbox can display JSON string, but subsequent changes to the "First" or "Last" text box will not change the value in the "Name" textbox.

Is it possible to bind an object directly? (I understand I can use JsObservable to observe the name but I wonder if I can do it with JsViews only.)

**Comment by johan-ohrn** [comment]
To explain what you're doing wrong just look at the following template:
```
<script id="myTmpl" type="text/x-jsrender">
  Name (as JSON): <input data-link="{toJson:name || name.first || name.last:fromJson}"/><br/><br/>

  First: <input data-link="name^first"/>{^{:name^first}}<br/>
  Last: <input data-link="name^last"/>{^{:name^last}}
</script>

<div id="page"></div>
```
There are a couple of issues with your code. Your fromJson converter creates and returns a new object. It does not modify properties on an already existing object. What this means is that when you type into the json textbox your converter will be called and it will return a completely new object. This is why you need to use the ```name^first``` expression for your first and last inputs. The ```^``` character makes the expression listen to modification of the name object itself.

Next issue is pretty much the same. Typing into the first or last inputs modifies those properties on the name object but your json input does not monitor those properties. It's looking to see if the name object itself is replaced by a new instance which it's not. A hackish solution I showed you just includes the first and last properties in the expression to trick the template into updating when the properties change as well but return the actual name object.

So basically you're not keeping track on which objects and properties are being monitored by each data link expression.

I wouldn't say this is a clean solution but I wanted to make you understand why you have these issues.

**Comment by hkvstore** [comment]
Thank you for your explanation. What I'm trying to do was to use JsViews to modify a JSON string with form controls. What will be the best solution if the object has a lot of properties and it becomes hard to put them all in the expression to trick the template?

**Comment by johan-ohrn** [comment]
You could use a computed observable instead of the converter.
Try this:
```
<script id="myTmpl" type="text/x-jsrender">
  Name (as JSON): <input data-link="json()"/><br/><br/>

  First: <input data-link="name.first"/>{^{:name.first}}<br/>
  Last: <input data-link="name.last"/>{^{:name.last}}
</script>

<div id="page"></div>
```

```
var myTmpl = $.templates("#myTmpl"),
  data = { name: { first: "Jo", last: "Blow" } };
  
  // create a computed observable
  data.json = function() {
  	return JSON.stringify(data.name);
  }
  data.json.set = function(val) {
    var x = JSON.parse(val);
    // observably update the individual properties on the name object
    $.observable(data.name).setProperty(x);
  }
  // make the computed observable depend on whatever properties are on the name object
  data.json.depends = ['name.first', 'name.last'];

myTmpl.link("#page", data);
```

**Comment by johan-ohrn** [comment]
@BorisMoore as discussed in this [issue](https://github.com/BorisMoore/jsviews/issues/417#issuecomment-480561407) it was stated that parens at the leaf node is optional for depends and observe paths. Does this apply to data-link expressions in templates as well?
This template work correctly ```<input data-link="json()"/>``` but ```<input data-link="json"/>``` does not.
Here's a [fiddle](http://jsfiddle.net/6wvtm8s2/1/). Type into the name textboxes and the second json textbox to see the effect.

**Comment by hkvstore** [comment]
> (I understand I can use JsObservable to observe the name but I wonder if I can do it with JsViews only.)

**Comment by johan-ohrn** [comment]
Use Object.keys(name) to get the list of property names to make it dynamic like so ```data.json.depends = Object.keys(data.name).map(function(x){return "name."+x});```

**Comment by hkvstore** [comment]
Or ``data.json.depends = ['name.*'];``

**Comment by johan-ohrn** [comment]
Hah how simple. I didn't even consider using '*' . Excelent!

**Comment by BorisMoore** [comment]
@johan-ohrn  and @hkvstore - thanks for your discussion above. 

Johan, in response to your question, above, parens are required, in the case of `data-link="someComputed()"`.

If you don't put the parens, on a `<span data-link="someComputed"`>` for example, then it renders as someComputed.toString(). However `<input data-link="someComputed"`>` will evaluate someComputed() because jQuery detects that it is a function, just in the case of `<input>` and evaluates it. So it is a spurious behavior...

    if ( isFunction ) {
        args[ 0 ] = value.call( this, index, self.html() );
    }

**Comment by BorisMoore** [comment]
I'll close this now...
